---
title: "ccr89_ffpe"
author: "Elsa"
date: "2022-12-07"
output: html_document
---
---
title: "Patient 89 analysis FFPE Samples"
output:
  html_document:
    keep_md: yes
    highlight: 'haddock'
    theme: cerulean
    df_print: paged
    number_sections: true
    toc: true
    toc_float: true
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    cache       = FALSE,     # if TRUE knitr will cache the results to reuse in future knits
    #fig.width   = 18,       # the width for plots created by code chunk
    #fig.height  = 8,       # the height for plots created by code chunk
    fig.align   = 'center', # how to align graphics in the final doc. 
    warning     = FALSE,
    cache       = FALSE,
    comment     = NA
)
# knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

The main purpose of this script is analyzing diversity measures as the relative abundance of microorganisms in paraffin samples from
patient 89 (subject_type : CCR89). In summary, the paper will focus on analyzing several input files, preprocessing them before and after importing as phyloseq 
object (before: using negative controls to remove contaminants; after: removing from samples a list of described contaminants from literature), extracting measures values and create bar plots for relative abundance in each type of sample. 

# Libraries

Loading the necessary libraries: 

```{r message=FALSE, warning=FALSE}
library(purrr)
library(ggplot2)
library(dplyr)
library(vegan)
library("devtools")
library(data.table)
library(ggtree)
library(phyloseq)
library(plotly)
library(parallel)
library(VennDiagram)
require(RColorBrewer)
library(egg)
library("qiime2R")
library(stringr) 
library(ggpubr)
library(glue)
library(randomcoloR)
library(glue)
library(DT)
options(dplyr.summarise.inform = FALSE) # remove summarise output message
print(glue("In case of problems with libraries, there is an installation instructions file, called installation_libraries_instruction.R, in the same folder of this script."))
```

# Visual functions

Load a function to show pretty tables in HTML format:
```{r}
create.table <- function(df, capt=NULL){
  DT::datatable(df, extensions = c('FixedColumns'),
                options = list(scrollX = TRUE, pageLength = 5), 
                caption=htmltools::tags$caption(
                  style = 'caption-side: bottom; text-align: center; color: black;',
                  htmltools::em(capt)
                ))
}
```

# Input
Here are specified the required inputs for this script (remember changing files paths):
```{r}
#######################
# ---- Load data ---- #
#######################
smurf="/home/elsa/Documentos/Doctorado/CCR89_junio22/SMURF/"
muestra="reconstruction_CCR89/"
ruta=paste(smurf, muestra, sep = "", collapse = NULL)
# metadata_file="metadataccr89_FFPE_controlesnegativos_nov22.tsv" 
#ruta_metadata =paste(smurf, metadata_file , sep = "", collapse = NULL)
# ruta_metadata = "/home/elsa/Documentos/Doctorado/CCR89_junio22/articulo_parvimonas_scripts/metadataccr89_FFPE.tsv"
ruta_metadata = "/home/elsa/microbiome_FFPE/data/formarted_ffpe_metadata.tsv"
taxonomia_file ="reconstructed-taxonomy-CCR-smurf-con-controles-neg-PE-CCR89_julio22.qza"
ruta_taxonomia =paste(ruta , taxonomia_file , sep = "", collapse = NULL)
# otus_file ="reconstructed-table-CCR-smurf-con-controles-neg-PE-CCR89_julio22.qza"
otus_file ="reconstructed-table-CCR89_julio22.qza"
ruta_otus =paste(ruta , otus_file , sep = "", collapse = NULL)
bact_contaminant_file = "/home/elsa/Documentos/Doctorado/CCR89_junio22/bacterias_contaminantes_unidas.tsv"


variables_of_interest <- c("id_tissue", "subject_type")
timeline_var <- ""
```

# Load data
QIIME2R library used to import qza files without exporting them to tsv or biom formats:

## Features

Load Amplicon Sequence Variants (ASVs) table and remove those which are empty (bad quality or a control): 
```{r}
seedcolors <- c("#ff0000", "#00ff00", "#0000ff")

############################
# ---- QIIME2R import ---- #
############################

# ---- ASVs -----
# Import directly from qiime2 format thanks to QIIME2R library
otu_table  <- read_qza(ruta_otus )
# head(otu_table )
otu  <- otu_table$data
# Transform to ASVs table to matrix
otu  = as.matrix(otu)
# rownames(otu)

otu_df <- otu 
cat("Nrows of ASVs table before checking for colSums == 0", nrow(otu_df))

# Remove columns which colSums would be equal to zero (no data)
condition <- colSums(otu_df) != 0
otu_df_empty_cols <- colnames(otu_df)[!condition]
if (length(otu_df_empty_cols)>0){
  print(glue("WARNING: There are {length(otu_df_empty_cols)} empty columns in ASV dataframe (no data, this can happen in controls) and they will be removed: '{paste(otu_df_empty_cols, collapse='; ')}'"))
}

otu_df <- otu_df[, condition]


```

(
## Taxonomy

Load taxonomy and propagate levels (filling NAs in taxa levels by the last known level):
```{r}
# ---- Taxmat ----

taxonomyTable  <-read_qza(ruta_taxonomia )
# head(taxonomyTable )
taxonomy  <- parse_taxonomy(taxonomyTable$data)
tax  = as.matrix(taxonomy )
class(tax )

# Propagate NAs
keep_unidentified <- function(tx_table){
  # Keep the ones that have one of c(NA, "", " ", "\t") in target_level column
  # and then find the last level that is known (can be Genus, Family...)
  # Assign to the target_level level the last known level + x__XXX NA
  # Family (last known)   Species
  # Lachnospiraceae    Lachnospiraceae NA
  elements_to_target <- colnames(tx_table)
  tx_table <- as.data.frame(tx_table)
  previous_level <- elements_to_target[1]
  for (level in elements_to_target[2:length(elements_to_target)]){
    empty_values <- tx_table[[level]] %in% c(NA, "", " ", "\t")
    tx_table[[level]][empty_values] <- paste(tx_table[[previous_level]][empty_values], "NA", sep="_")
    tx_table[[level]][empty_values] <- gsub("(_NA[> ]*)*", "\\1", tx_table[[level]][empty_values]) # Replace repetitions of NA by one NA
    
    previous_level <- level
  }
  
  return(tx_table)
}

taxmat <- as.matrix(keep_unidentified(taxonomy))

# Set row names
# rownames(taxmat) <- rownames(otu_df)

```


## Metadata

Processing Metadata file:
```{r}
# ---- Metadata ----
sample_data  <- is_q2metadata(ruta_metadata )
sample_data  <- read_q2metadata(ruta_metadata )
rownames(sample_data) <- sample_data[,'SampleID']

# Check if there are columns not in otu_df besides the otu_df columns that were empty
condition <- sample_data[,'SampleID'] %in% colnames(otu_df)
SampleIDs.not.in.otus <- sample_data[!condition,]$SampleID
sample_data <- sample_data[condition,] # removing samples from sample_data not presented in otu_df
rownames(sample_data) <- sample_data[,'SampleID']

# Create column for grouping by variables_of_interest
if ("grouping_var" %in% names(sample_data)){
  stop("Ups, sample_data already has grouping_var as column")
} else {
  sample_data$grouping_var <- do.call(paste, c(sample_data[variables_of_interest], sep="__"))
}

# Make categorical columns
for (i in variables_of_interest){
  sample_data[[i]] <- factor(sample_data[[i]])
}


```

# Apply to controls

Subtract controls specified in metadata (sample_data), in order to reduce possible PCR or paraffin or human manipulating samples contaminates:
```{r}
# ---- Controls ----
# Calculate each ASV sum in controls
sum_rows <- function(input){
  # If input is a Data Frame, obtain vector with rowSums 
  # If it is a vector, return 
  if (is.null(ncol(input))){
    return (input)
  } else {
    return (rowSums(input))
  }
}


substract_control <- function(features, sample_data){
  cat("\nUsing controls for cleaning...\n")
  
  # - Debug-
  #features = otu_df
  
  # Initialize clean dataframe with non-control ids
  clean.otu.df <- features[, subset(sample_data, sample_nature!="control")$SampleID]
  
  # Split "control_applies_to" column by ";"
  # The resulting elements will be searched in "id_tissue" column 
  # and the control will be applied to corresponding samples
  sample_data$control_applies_to_split <- strsplit(as.character(sample_data$control_applies_to), ";")
  # For each control ID
  for (control_id in subset(sample_data, sample_nature=="control")$SampleID ){
    # Get the sample_data of the control ID
    r <- subset(sample_data, SampleID==control_id)
    # Log
    control_id_seq_run = r$sequencing_run
    print(glue("\n==== Using SampleID '{control_id}' as control in sequencing_run='{control_id_seq_run}' ===="))
    
    # Check that if element "all" is present no other elements exist
    # and that an element is not repeated with unique
    applies_to_elements <- unique(r$control_applies_to_split[[1]]) # To access split elements: sample_data$control_applies_to_split[65][[1]][1]
    if ("all" %in% applies_to_elements & length(applies_to_elements)>1){
      stop("Element 'all' was specified, but extra elements are present. Please choose either 'all' or the other elements")
    }
    # For each id_tissue to which the control applies to
    for (applies_to in applies_to_elements){ 
      print(glue("\n~~~ Applying control to: '{applies_to}' ~~~"))
      tmp <- subset(sample_data, sample_nature!="control" & sequencing_run==control_id_seq_run)
      # If the control applies to all samples
      if (applies_to == "all"){
        tmp <- tmp
        # If it only applies to a certain id_tissue type which IS in sample_data$id_tissue
      } else if (applies_to %in% sample_data$id_tissue){
        tmp <- subset(tmp, id_tissue==applies_to)
        # If it is not "all" or the element does not exist in sample_data$id_tissue, raise error 
      } else {
        stop(glue("id_tissue '{applies_to}' is not in sample_data$id_tissue column, please fix"))
      }
    
      if (length(tmp$SampleID) >= 1){
        print(glue("Applying control.id={control_id} from sequencing_run={control_id_seq_run} to element={applies_to} in {length(tmp$SampleID)} samples"), "\n")
        if (length(tmp$SampleID) >= 8){
          t <- as.character(tmp$SampleID)
          clean.otu.df[, t] <- clean.otu.df[, t] - sum_rows(features[,c(control_id)]) 
        } else{
          t <- tmp$SampleID
          clean.otu.df[, t] <- clean.otu.df[, t] - sum_rows(features[,c(control_id)])  
        }
       } else {
        print(glue("WARNING: Control control.id={control_id} from sequencing_run={control_id_seq_run} cant be applied to element={applies_to}, no samples match ({length(tmp$SampleID)} samples)\nThis in unexpected, every control should apply to at least one sample! Check sample_data? Stopping..."))
        # 
        if (control_id_seq_run == "v2smurf11062021" | control_id_seq_run == "v2smurf08102021"){
          print(glue("Applying control.id={control_id} from sequencing_run={control_id_seq_run} to element={applies_to} in not matching samples."), "\n")
          not.sequencing.run.match.samples <- subset(sample_data, SampleID %in% c("CCR-89-1-FFPE-Tumor", "CCR-89-1-FFPE-Normal"))
          clean.otu.df[, not.sequencing.run.match.samples$SampleID] <- clean.otu.df[, not.sequencing.run.match.samples$SampleID] - sum_rows(features[,c(control_id)])
        }
      }
    }
  }
  # Negative numbers to 0
  clean.otu.df[clean.otu.df<0] <- 0
  
  return(clean.otu.df)
}
  
clean.otu.df <- substract_control(otu_df, sample_data)

```


# Create phyloseq object

After preprocessing, the phyloseq object is created:
```{r}
# ---- Phyloseq object ----
OTU = otu_table(clean.otu.df, taxa_are_rows = TRUE)
TAX = tax_table(taxmat)
SAM = sample_data(sample_data)

physeq = phyloseq(OTU, TAX, SAM)
physeq

# saved RData
# smurf_ccr89 <- save(OTU, TAX, SAM, physeq, file = "smurf_ccr89.RData")
```

# Remove typical contamination by name

As an additional removing contaminant step, we remove possible contaminant ASVs using their genus and known contaminants from the input "bact_contaminant_file":
(NOTE: during the processing, there will be adding two additional contamination microorganism to bact_contaminant_file directly in R)

```{r}
# ---- Apply simple contamination filter using genus ----
pop_taxa = function(ps, bad_asvs){
  all_asvs = taxa_names(ps)
  keep_these <- all_asvs[!(all_asvs %in% bad_asvs)]
  return(prune_taxa(keep_these, ps))
}
# Only bacteria
bad_asvs <- taxa_names(subset_taxa(tax_table(physeq), Kingdom!="d__Bacteria"))
physeq <- pop_taxa(physeq, bad_asvs)

# Remove typical contamination
typical_contamination <- read.csv(bact_contaminant_file, sep='\t')
# typical_contamination$Genus <- paste("g__", typical_contamination$Genus, sep="") # not with qiime2R processing
# adding new file to df_bacterias for Bulkholderia Genus
typical_contamination[nrow(typical_contamination) + 1,] = c("Bacteria", "Proteobacteria", "Betaproteobacteria","Burkholderia-Caballeronia-Paraburkholderia")
# adding new contaminant bacteria: Dickeya
typical_contamination[nrow(typical_contamination) + 1,] = c("Bacteria", "Proteobacteria", "Gammaproteobacteria","Dickeya")


cont_genus <- c(typical_contamination$Genus, paste(typical_contamination$Genus, "NA", sep= "_"))
physeq <- subset_taxa(physeq, !Genus %in% cont_genus)

# Manual names
bad_asvs <- taxa_names(subset_taxa(tax_table(physeq), Family %in% c("Dermacoccaceae", "Dermacoccaceae|Micrococcaceae","Mitochondria","Chloroplast")))
physeq <- pop_taxa(physeq, bad_asvs)

# Keep ASVs with more than one count
physeq <- prune_taxa(taxa_sums(physeq) > 1, physeq) 

# Show change
physeq
```


# Proportions and barplots

Relative abundance can now be calculated.

## Some functions

Before any calculation, we describe the functions that will be used in these steps:

```{r}
##############################################################################
# ------ Abundance+Prevalence filters and relative abundance functions ----- #
##############################################################################

get.otus.to.keep.by.prevalence <- function(group, data, abundance_threshold, prevalence_threshold){
  starting.otus.count <- length(unique(data$OTU))
  
  print(glue("\nFiltering group composed of: '{paste(group, collapse='; ')}'"))
  
  # For each sample, filter to OTUs that have a relative abundance over N% 
  # and appear in at least M% of the other samples in its group
  abundance_over_list <- list()
  abundance_under_list <- list()
  for (sample_id in group){
    # For each SampleID in group
    # Calculate relative abundance for the SampleID's data
    # Get OTUs over threshold into a list
    df <- subset(data, SampleID==sample_id)
    df$relative_abundance <- df$Abundance/sum(df$Abundance)
    otus_over_thresh <- subset(df, relative_abundance>=abundance_threshold)$OTU
    otus_under_thresh <- subset(df, relative_abundance<abundance_threshold)$OTU
    abundance_over_list[[sample_id]] <- otus_over_thresh
    abundance_under_list[[sample_id]] <- otus_under_thresh
    print(glue("Sample {sample_id} keeps {length(otus_over_thresh)}/{length(otus_under_thresh)+length(otus_over_thresh)} ASVs filtering by abundance={abundance_threshold}"))
  }
  
  # Transform list of vectors into a presence/absence dataframe by matching ASV names
  # leaving NAs if a sample does not have the ASV.
  # Each row is an ASV, each column is a sample
  # Example:
  #                             ccr11-1s                       CCR-59-1-s                       CCR-57-1-s
  # 1   101637aba474a6614786486ad42acb11 101637aba474a6614786486ad42acb11 101637aba474a6614786486ad42acb11
  # 2   00a2f7e2564becc62e71c46821779802                             <NA>                             <NA>
  # 3   f7cf33648c7d6bda19e991b97583cdab f7cf33648c7d6bda19e991b97583cdab f7cf33648c7d6bda19e991b97583cdab
  # 21  0a49a32752b611a144ddd506a2155134                             <NA> 0a49a32752b611a144ddd506a2155134
  # 23  c773422b3dd3b7f057a6ec0d5fb93733 c773422b3dd3b7f057a6ec0d5fb93733                             <NA>
  otu_presence <- abundance_over_list %>%
    map(~ data.frame(col = ., ., stringsAsFactors = FALSE)) %>%
    reduce(full_join, by = "col") %>%
    select(-col) %>%
    setNames(names(abundance_over_list))

  # Set the OTUs as row names
  rownames(otu_presence) <- apply(otu_presence, 1, function(z) na.omit(z)[1])
  
  # Select which OTUs have to be kept by prevalence
  # The ASVs that appear in at least X% of the samples (row-wise)
  condition <- (rowSums(!is.na(otu_presence)) / dim(otu_presence)[2]) >= prevalence_threshold
  otus_to_keep <- rownames(otu_presence)[condition]
  otus_to_loose_by_prevalence <- rownames(otu_presence)[!condition]
  print(glue("Surviving ASVs to abundance filter kept by prevalence: {length(otus_to_keep)}/{length(otus_to_loose_by_prevalence)+length(otus_to_keep)}"))
  
  print(glue("Final: {length(otus_to_keep)}/{starting.otus.count} ASVs have passed the abundance+prevalence filter."))
  return(
    list(otus_to_keep=otus_to_keep, otus_to_loose_by_prevalence=otus_to_loose_by_prevalence, 
         abundance_over_list=abundance_over_list, abundance_under_list=abundance_under_list
         )
    )
}

group_otus <- function(physeq_long, variables_of_interest, elements_to_target){
  # Group OTUs by ID, variables_of_interest, subject and elements_to_target in case 
  # there are repeated OTUs
  tmp_cols <- c("OTU", variables_of_interest, "SampleID", elements_to_target)
  before_grouping_count <- dim(physeq_long)
  physeq_long <- physeq_long %>%  
    dplyr::group_by(across(all_of(tmp_cols))) %>% 
    dplyr::summarise(
      Abundance = sum(Abundance), # sum the abundances in the group
      n = n() # how many rows existed in the group (Important for testing)
    )
  after_grouping_count <- dim(physeq_long)
  print(glue("Reduction of data rows from {before_grouping_count[1]} to {after_grouping_count[1]} by grouping with --> {paste(tmp_cols, collapse='; ')}"))
  
  return(physeq_long)
}

calculate_props <- function(physeq_long, variables_of_interest, target_level, threshold){
  # ------------------------------------- #
  # Calculate relative abundance per group
  # - physeq_long is expected to be a grouped dplyr dataframe
  # ------------------------------------- #
  
  # - Debug -
  # physeq_long <- aggregated_samples

  # Replace .*__ with an emtpy string in target column
  physeq_long[[target_level]] <- gsub(".*__", '', physeq_long[[target_level]])
  
  # Convert to data.table
  physeq_long <- data.table(physeq_long)
  
  # Group by OTU, target_level, Type and Site
  grouped_df <- physeq_long[, mean(Abundance), by = c('OTU', target_level, variables_of_interest)]
  colnames(grouped_df) <- c('OTU', target_level, variables_of_interest, 'Abundance')
  
  # Relative abundance
  grouped_df$relative_abundance <- grouped_df$Abundance/sum(grouped_df$Abundance)
  
  # Reorder by abundance
  grouped_df <- grouped_df[order(grouped_df[,'relative_abundance'], decreasing=TRUE),]
  
  
  # Merge rows under a certain abundance threshold to "other"
  under_thresh_df <- grouped_df[grouped_df$relative_abundance < threshold,]
  over_thresh_df <- grouped_df[grouped_df$relative_abundance >= threshold,]
  
  # WARNING: expects only two variables of interest ("var1" and "var2")
  under_thresh_df <- data.frame(t(c('__OTHER__', '__Other__', "var1", "var2", sum(under_thresh_df$relative_abundance), sum(under_thresh_df$Abundance)))) # FIXME: error here if using timeline
  names(under_thresh_df) <- c('OTU', target_level, variables_of_interest, 'relative_abundance', "Abundance")
  
  final_df <- rbind(over_thresh_df, under_thresh_df)
  final_df$relative_abundance <- as.numeric(final_df$relative_abundance)
  final_df$Abundance <- as.numeric(final_df$Abundance)
  
  # Select columns (return value cannot include grouping variables in group_modify)
  return(data.frame(final_df)[,c("OTU", target_level, 'relative_abundance', "Abundance")])
}

physeq_proccessing <- function(physeq_obj, target_level, elements_to_target, filter_abundance_threshold, 
         prevalence_threshold, plot_abundance_threshold, sample_data, out_folder, variables_of_interest){
  # - Debug -
  # physeq_obj = physeq
  # target_level = "Genus"
  # elements_to_target = c("Kingdom","Phylum", "Class", "Order", "Family", "Genus")
  # filter_abundance_threshold = 0.000001
  # prevalence_threshold = 0.2
  # plot_abundance_threshold = 0.000001
  # threshold = plot_abundance_threshold
  # sample_data = sample_data
  # out_folder = ""
  # variables_of_interest = variables_of_interest
  
  # Check
  if (plot_abundance_threshold<filter_abundance_threshold){
    stop("Plot abundance threshold can not be lower than the abundance threshold. It could lead to bad interpretation of the results. Stopping...")
  }
  asv_count_per_step <- list()
  
  # Melt physeq
  print(glue("Melting phyloseq object to long format..."))
  data_phylo_long <- physeq_obj %>%
    tax_glom(taxrank = target_level, bad_empty = c()) %>%   # Agglomerate to a specific taxonomic rank 
    psmelt()                                                # Change to "long" format 
  asv_count_per_step[["starting_point"]] = dim(data_phylo_long)[1]
  
  #####################
  # ---- Filters ---- #
  #####################
  # ---- Filter by abundance and prevalence ----
  # Filter to OTUs that have a relative abundance over N% 
  # and appear in at least M% of the group's samples
  print(glue("Prevalence and abundance based filtering..."))
  
  tmp <- subset(sample_data, sample_nature!="control") %>%
    group_by(across(all_of(variables_of_interest))) %>%
    group_split()

  groups <- lapply(tmp, function(x){x$SampleID})
  for (group in groups){
    cat("Group:", group, "\n")
    # Get OTUs to keep by prevalence
    tmp_res <- get.otus.to.keep.by.prevalence(group, data_phylo_long, filter_abundance_threshold, prevalence_threshold)
    otus_to_keep <- tmp_res$otus_to_keep

    # Remove
    data_phylo_long <- data_phylo_long %>% filter(!(SampleID %in% group & !OTU %in% otus_to_keep))
    cat("\n")
  }
  asv_count_per_step[["abundance_and_prevalence"]] = dim(data_phylo_long)[1]
  
  # ---- Grouping same OTUs ----
  print(glue("Group OTUs by ID and variables of interest and sum abundance..."))
  data_phylo_long <- group_otus(data_phylo_long, variables_of_interest, elements_to_target)
  asv_count_per_step[["group_by_asv"]] = dim(data_phylo_long)[1]
  
  ##################################
  # ---- Defining Replicates ----#
  ##################################
  print(glue("Merging replicated samples and obtaining their Abundance..."))
  # ---- Renaming and merging replicates ---
  naming_replicates <- list(
    c("CCR-89-1-FFPE-Normal", "CCR-89-1-FFPE-Normal-b"),
    c("CCR-89-1-FFPE-Normal", "CCR-89-1-FFPE-Normal-c"),
    c("CCR-89-1-FFPE-Tumor", "CCR-89-1-FFPE-Tumor-a"),
    c("CCR-89-1-FFPE-Tumor", "CCR-89-1-FFPE-Tumor-c"), 
    c("CCR_FFPE_89_1_tumor_2", "CCR-FFPE-89-1-tumor-2-f"),  
    c("CCR_FFPE_89_1_adenoma_1", "CCR-FFPE-89-1-adenoma-1-e" ),
    c("CCR_FFPE_89_1_adenoma_2","CCR-FFPE-89-1-adenoma-2-d"), 
    c("CCR-FFPE-89-1-adenoma-3", "CCR-FFPE-89-1-adenoma-3-c"),
    c("CCR-FFPE-89-1-adenoma-3", "CCR_FFPE_89_1_adenoma_3g")
    )
  
  d_p_l <- data_phylo_long
  d_p_l$num_repl <- 1 # number of replicates
  d_p_l$replicates <- d_p_l$SampleID
  for (i in naming_replicates){
    #print(i)
    d_p_l$replicates <- gsub(i[2], i[1], d_p_l$replicates)
    d_p_l$num_repl[d_p_l$replicates == i[1]] <- d_p_l$num_repl[d_p_l$replicates == i[1]] + 1
  }
  # ---- Real Abundance ----
  d_p_l$Abundance <- d_p_l$Abundance / d_p_l$num_repl
  
  # ---- Merging replicates samples and sum their abundances ----
  aggtmp_cols <- c("OTU", "id_tissue","replicates", "SampleID", "subject_type", elements_to_target, target_level)
  aggregated_samples <- d_p_l %>% 
    dplyr::group_by(across(all_of(aggtmp_cols))) %>% 
    dplyr::summarise(Abundance = sum(Abundance))

  ###################################
  # ---- Calculate proportions ---- #
  ###################################
  
  # ---- Proportions for each group ----
  print(glue("Calculate relative abundance by group..."))
  #variables_of_interest2 = c("replicates","id_tissue")
  print(target_level)
  proportions_per_group <- aggregated_samples %>%  
    dplyr::group_by(id_tissue) %>%
    dplyr::group_modify(~calculate_props(.x, variables_of_interest, target_level, plot_abundance_threshold), .keep=TRUE)
  
  # Remove empty classes
  proportions_per_group <- subset(proportions_per_group, Abundance != 0)
  asv_count_per_step[["final_proportions"]] = dim(data_phylo_long)[1]
  
  # Write cleaned long data
  write.table(data_phylo_long, file = glue("{out_folder}/cleaned_long.tsv"), row.names = FALSE, sep = "\t")
  
  # Write count log
  count_log <- as.data.frame(asv_count_per_step)
  count_log$filter_abundance_threshold <- filter_abundance_threshold
  count_log$prevalence_threshold <- prevalence_threshold
  count_log$plot_abundance_threshold <- plot_abundance_threshold
  write.table(count_log, file = glue("{out_folder}/count_log.tsv"), row.names = FALSE, sep = "\t")
  
  return(list(proportions=proportions_per_group, cleaned_long_data=data_phylo_long))
}


####################################
# ------- Bar plot function ------ #
####################################
get_barplots <- function(proportions, target_level, y="relative_abundance", title="", column_order=NULL){
  # Colors
  color <- grDevices::colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)] # Creando vector de colores contrastantes
  phylum_colors <- distinctColorPalette(length(levels(as.factor(proportions[[target_level]]))))
  
  # Change color of __Other__ to black
  others_pos <- levels(as.factor(proportions[[target_level]])) == "__Other__"
  phylum_colors[others_pos] <- "black"
  
  # Split names by '_'
  proportions[[target_level]] <- gsub("_", ' ', proportions[[target_level]])
  
  # Config plot
  geom_bar_pos <- NULL
  ylab <- NULL
  if (y=="relative_abundance"){
    geom_bar_pos <- "fill"
    ylab <- "Relative Abundance (RA) %"
  } else if (y=="Abundance"){
    geom_bar_pos <- "stack"
    ylab <- "Abundance"
  }
  
  # Main
  p <- ggplot(proportions, aes_string(x = "label", y = y, fill = target_level)) + 
    geom_bar(stat = "identity", position = geom_bar_pos) + 
    # facet_wrap(~ origin , scales="free_x") +
    scale_fill_manual(values = phylum_colors)
  
  if (y=="relative_abundance"){
    p <- p + scale_y_continuous(labels = scales::percent_format())
  }
  if (!is.null(column_order)){
    p <- p + scale_x_discrete(limits = column_order)
  }
  
  p <- p + guides(fill = guide_legend(reverse = TRUE, keywidth = 1, keyheight = 1)) + 
    ylab(ylab) + ggtitle(title) + xlab("") + theme_classic() + 
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5)) + 
    theme(axis.text.x=element_text(angle = 90), legend.key.size = unit(0.5,"line")) +
    guides(fill=guide_legend(ncol=1, title=target_level)) # + theme(legend.position="none")
  
  return(list(barplot=p))
  
}


```

## Calculate and plot

Now, we can calculate proportions of each sample type (in this paper we only analyze three of them: Adenoma (A), Adenocarcinoma (AC) or Normal-Colon (NC) at each taxonomic rank (Family, Genus, Species, ...) with specific abundance and prevalence thresholds.

```{r}
 ###################################
 # ----- Creating RA bar plots ---- #
 ###################################

# - Debug -
# target_level <- "Genus" # Ex: "Genus"
# elements_to_target <- c("Kingdom","Phylum", "Class", "Order", "Family", "Genus") # Ex: c("Kingdom","Phylum", "Class", "Order", "Family", "Genus")

# ---- For paper results ----
# filter_abundance_threshold <- 0.000001
# prevalence_threshold <- 0.2
# plot_abundance_threshold <- 0.000001


# ---- For plotting ----
filter_abundance_threshold <- 0.000001
prevalence_threshold <- 0.2
plot_abundance_threshold <- 0.01


# Get proportions
conf = list(
 list("Family", c("Kingdom","Phylum", "Class", "Order", "Family")),
 list("Genus", c("Kingdom","Phylum", "Class", "Order", "Family", "Genus"))
)

result_list <- list()
for (i in conf){
 target_level <- i[[1]]
 elements_to_target <- i[[2]]
 out_folder = glue("abund-{filter_abundance_threshold}__prev-{prevalence_threshold}__plot-{plot_abundance_threshold}/{target_level}")
 dir.create(out_folder, showWarnings = FALSE, recursive = TRUE)

 # Get proportions
 prop <- physeq_proccessing(
   physeq_obj=physeq, target_level=target_level,
   elements_to_target=elements_to_target,
   filter_abundance_threshold=filter_abundance_threshold, prevalence_threshold=prevalence_threshold,
   plot_abundance_threshold=plot_abundance_threshold, sample_data=sample_data, out_folder=out_folder, variables_of_interest=variables_of_interest)

 # Create label
 # prop$proportions$label <- paste(prop$proportions$replicates, prop$proportions$id_tissue, sep='__')
 # prop$proportions$label <- paste(prop$proportions$id_tissue, sep='__')
 prop$proportions$label <- prop$proportions$id_tissue

   # Rename a tag
 renaming_list <- list(
   c("Normal-colon", 'M89-PT-NC'),
   c("Adenoma", 'M89-PT-A'),
   c("Tumor", 'M89-PT-Ac')
 )

 for (i in renaming_list){
   print(i)
   prop$proportions$label <- gsub(i[1], i[2], prop$proportions$label)
 }
   # ---- Group samples by tissue type and extract mean as values and also as percentage ----
  # Although we have not analyzed Species rank level in this type of samples, we let this in code.
  if (target_level == "Family"){
    z <-  group_by(prop$proportions, label, Family) %>% summarise(relative_abundance = mean(relative_abundance))
    z$mean_real_ab_per <- z$relative_abundance*100

  } else if(target_level == "Genus"){
    z <- group_by(prop$proportions,label, Genus) %>% summarise(relative_abundance = mean(relative_abundance))
    z$mean_real_ab_per <- z$relative_abundance*100

  } else{
    z <-  group_by(prop$proportions, label, Species) %>% summarise(relative_abundance = mean(relative_abundance))
    z$mean_real_ab_per <- z$relative_abundance*100
  }
  column_order <- c("M89-PT-NC", "M89-PT-A", "M89-PT-Ac") 
  z <- z[z$label %in% column_order,] # to only get the chosen labels
  if (dim(z)[1]>=1){ # if data frame contains any data --> save
        file_name <- glue("{out_folder}/{target_level}__mean_relat_abund.tsv")
        print(glue("Saving data  to {file_name}..."))
        write.table(z, file = file_name, row.names = FALSE, sep = "\t"
        )
    }
   
   barplot <- get_barplots(
     proportions=z,
     target_level=target_level,
     title=glue("{target_level} (Abundance:{filter_abundance_threshold}, Prevalence:{prevalence_threshold}, Plot:{plot_abundance_threshold})"),
     column_order=column_order
     )
 
   # Save plot
   # htmlwidgets::saveWidget(
   #   as_widget(ggplotly(barplot$barplot)),
   #   glue("{out_folder}/barplot.html")
   # )
 
   result_list[[target_level]] <- list(proportions=prop$proportions, barplot=barplot$barplot)
 }




```
# 
```{r}
print(result_list$Family$barplot)
print(result_list$Genus$barplot)
```

# For an interactive view
```{r}
ggplotly(result_list$Family$barplot)
ggplotly(result_list$Genus$barplot)
```

```{r message=FALSE, warning=FALSE}
# Save images
tiff(paste("parvimonas_16S_89_t_ffpe", plot_abundance_threshold, ".tiff", sep=''), width=15,height=15, units = 'in', res =300)
ggarrange(result_list$Family$barplot, result_list$Genus$barplot, ncol = 2)
dev.off()
```

# With Species

Add in conf list variable:  list("Species", c("Kingdom","Phylum", "Class", "Order", "Family", "Genus", "Species")) (remember adding commas after the last element in previous conf list).

<!-- ```{r} -->
<!-- print(result_list$Family$barplot) -->
<!-- print(result_list$Genus$barplot) -->
<!-- print(result_list$Species$barplot) -->
<!-- ``` -->

<!-- ```{r message=FALSE, warning=FALSE} -->
<!-- # Save images -->
<!-- tiff(paste("parvimonas_16S_89_t_ffpe", plot_abundance_threshold, ".tiff", sep=''), width=40,height=30, units = 'in', res =300) -->
<!-- ggarrange(result_list$Family$barplot, result_list$Genus$barplot, result_list$Species$barplot, ncol = 2, nrow=2) -->
<!-- dev.off() -->
<!-- ``` -->

<!-- ```{r} -->
<!-- ggplotly(result_list$Family$barplot) -->
<!-- ggplotly(result_list$Genus$barplot) -->
<!-- ggplotly(result_list$Species$barplot) -->
<!-- ``` -->

# Session info

Session info (including used packages), is available in the following code section:

```{r}
sessionInfo()
```
